File: leetcode/hasPathSum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {boolean}
 */
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}
var hasPathSum = function (root, sum) {
  if (!root) return false;
  if (root.val === sum) {
    if (root.left || root.right)
      return hasPathSum(root.left || root.right, sum - root.val);
    return true;
  }
  return (
    hasPathSum(root.left, sum - root.val) ||
    hasPathSum(root.right, sum - root.val)
  );
};
var hasPathSum = function (root, sum) {
  if (!root) return false;
  if (!root.left && !root.right && root.val === sum) return true;
  return (
    hasPathSum(root.left, sum - root.val) ||
    hasPathSum(root.right, sum - root.val)
  );
};
hasPathSum(
  new TreeNode(
    5,
    new TreeNode(4, new TreeNode(11, new TreeNode(7), new TreeNode(2))),
    new TreeNode(8, new TreeNode(13), new TreeNode(4, null, new TreeNode(1)))
  ),
  22
);

File: leetcode/hasPathSum.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def hasPathSum(
            self, root: Optional[TreeNode],
            targetSum: int) -> bool:
        return has_path_sum(root, targetSum)


def has_path_sum(root, targetSum):
    def dfs(node, cur):
        if not node:
            return False
        if not node.left and not node.right and cur == node.val:
            return True
        return dfs(node.left, cur - node.val) or dfs(node.right, cur - node.val)

    return dfs(root, targetSum)


has_path_sum(TreeNode(1, TreeNode(2)), 1)
